### PR TITLE
Implemented Create Task Feature.

### PR DESCRIPTION
**Pull request by mustafahasankhan**
### Proposed Changes

- Completed Create Task feature which gives user ability to create a new task.
- Completed Access Token fetching and implementation in API calls.

https://user-images.githubusercontent.com/65130881/176226086-b546b86c-aaa4-4d1a-a83c-90614a7be89a.mp4

### CODE REVIEW

File: src/handlers/ExecuteBlockActionHandler.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        private readonly read: IRead,
+        private readonly http: IHttp,
+        private readonly modify: IModify,
+        private readonly persistence: IPersistence
+    ) {}
+
+    public async run(
+        context: UIKitBlockInteractionContext
+    ): Promise<IUIKitResponse> {
+        console.log("reaching block action handler");
```
Comment (AllanPazRibeiro) [inline src/handlers/ExecuteBlockActionHandler.ts]: Is not a good practice to use console.log, I recommend adding app.getLogger.log instead or removing this log

File: src/handlers/ExecuteBlockActionHandler.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        const data = context.getInteractionData();
+        const { actionId, user } = data;
+
+        switch (actionId) {
+            case "trial": {
```
Comment (AllanPazRibeiro) [inline src/handlers/ExecuteBlockActionHandler.ts]: I don't know how many cases this switch will have, but it's better to create an enum to identify it

File: src/handlers/ExecuteBlockActionHandler.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+        try {
+			this.app.getLogger().info(data);
+            // await postTask({context,data,room,read,persistence,modify,http});
```
Comment (AllanPazRibeiro) [inline src/handlers/ExecuteBlockActionHandler.ts]: is this commented code necessary?

File: src/lib/persistence.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import { IPersistence, IPersistenceRead } from '@rocket.chat/apps-engine/definition/accessors';
+import { RocketChatAssociationModel, RocketChatAssociationRecord } from '@rocket.chat/apps-engine/definition/metadata';
+
+export const persistUIData = async (persistence: IPersistence, id: string, data: any): Promise<void> => {
```
Comment (AllanPazRibeiro) [inline src/lib/persistence.ts:4]: avoid using `any` as much as you can, I know in some circumstances is acceptable but some not

File: src/lib/postTask.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+}: {
+    context: UIKitViewSubmitInteractionContext,
+    data: IUIKitViewSubmitIncomingInteraction,
+    room: IRoom;
+    read: IRead;
+    persistence: IPersistence;
+    modify: IModify;
+    http: IHttp;
+}) {
+    // const { state }: ICreateTaskState = data.view as any;
```
Comment (AllanPazRibeiro) [inline src/lib/postTask.ts]: unnecessary commented code here

File: src/lib/postTask.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    const state = data.view.state;
+    const user: IUser = context.getInteractionData().user;
+    const token = await getAccessTokenForUser(read, user);
+    const list_id = state?.[ModalsEnum.LIST_ID_BLOCK]?.[ModalsEnum.LIST_ID_INPUT];
+    const taskName = state?.[ModalsEnum.TASK_NAME_BLOCK]?.[ModalsEnum.TASK_NAME_INPUT];
+    const taskDescription = state?.[ModalsEnum.TASK_DESCRIPTION_BLOCK]?.[ModalsEnum.TASK_DESCRIPTION_INPUT];
+    const headers: any = {
```
Comment (AllanPazRibeiro) [inline src/lib/postTask.ts]: I don't think that those any(s) are necessary

File: src/lib/postTask.ts [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        if (room) {
+            textSender.setRoom(room);
+        }
+    await modify.getCreator().finish(textSender);
+    }
+    else {
+        const textSender = await modify
+        .getCreator()
+        .startMessage()
+        .setText(`❗️ Unable to create task! \n Error ${response.data.err}`);
```
Comment (AllanPazRibeiro) [inline src/lib/postTask.ts:68]: Did this emoji work on the message?

File: src/lib/getTasks.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+}: {
+    context: UIKitBlockInteractionContext,
+    room: IRoom;
+    read: IRead;
+    persistence: IPersistence;
+    modify: IModify;
+    http: IHttp;
+}) {
+    
+    const api_response = await http.post(
```
Comment (AllanPazRibeiro) [inline src/lib/getTasks.ts]: I think the right HTTP verb should be `GET`, is it not?

File: src/lib/getTasks.ts [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+        `https://api.clickup.com/api/v2/list/tasks`
+    );
+    if(api_response.statusCode == 200){
+        const textSender = modify
+        .getCreator()
+        .startMessage()
+        .setText(`*TASK CREATED SUCCESSFULLY!*`);
```
Comment (AllanPazRibeiro) [inline src/lib/getTasks.ts]: I think you may consider separating the logic of sending messages in a module or something, and why send a message like this when getting the list of all tasks?

File: src/lib/getTasks.ts [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+        if (room) {
+            textSender.setRoom(room);
+        }
+        await modify.getCreator().finish(textSender);
+    }
+    else {
+        const textSender = modify
+        .getCreator()
+        .startMessage()
+        .setText(`*UNABLE TO CREATE TASK!*`);
```
Comment (AllanPazRibeiro) [inline src/lib/getTasks.ts]: The text should be something like *UNABLE TO GET TASKS!*

File: src/handlers/ExecuteViewSubmitHandler.ts [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+	constructor(
+		private readonly app: ClickUpApp,
+		private readonly read: IRead,
+		private readonly http: IHttp,
+		private readonly modify: IModify,
+		private readonly persistence: IPersistence
+        
+	) {}
+
+	public async run(context: UIKitViewSubmitInteractionContext,read: IRead,
```
Comment (AllanPazRibeiro) [inline src/handlers/ExecuteViewSubmitHandler.ts:21]: Maybe you may consider encapsulating the whole method in a try-catch, so there's will no need for multiple try-catches along with every case on the switch

File: ClickUpApp.ts [hunk 12]
```diff
@@ -105,6 +107,16 @@ export class ClickUpApp extends App {
...
     }
 
+    public async executeBlockActionHandler(context: UIKitBlockInteractionContext, read: IRead, http: IHttp, persistence: IPersistence, modify: IModify): Promise<IUIKitResponse> {
+        const handler = new ExecuteBlockActionHandler(this, read, http, modify, persistence);
+        return await handler.run(context);
+    }
+
+    public async executeViewSubmitHandler(context: UIKitViewSubmitInteractionContext, read: IRead, http: IHttp, persistence: IPersistence, modify: IModify) {
+        const handler = new ExecuteViewSubmitHandler(this, read, http, modify, persistence);
+        return await handler.run(context,read, http, persistence, modify);
```
Comment (AllanPazRibeiro) [inline ClickUpApp.ts]:
```suggestion
        return await handler.run(context, read, http, persistence, modify);
```

File: ClickUpApp.ts [hunk 13]
```diff
@@ -28,7 +30,7 @@ export class ClickUpApp extends App {
         super(info, logger, accessors);
     }
 
-    private oauth2ClientInstance: IOAuth2Client;
+    public oauth2ClientInstance: IOAuth2Client;
```
Comment (AllanPazRibeiro) [inline ClickUpApp.ts]: Why did you change this to `public` from `private`?

Maybe you should consider creating a public get method to expose this private prop, something like:
```

private oauth2Config......

public getOauth2Config() {
  return this.oauth2Config;
}
```

File: src/lib/postTask.ts (hunk 5 above)
Comment (mustafahasankhan) [reply src/lib/postTask.ts]: Sir, I referenced it here so that I may remember it later, it It is to be implemented this week.

File: src/lib/getTasks.ts (hunk 9 above)
Comment (mustafahasankhan) [reply src/lib/getTasks.ts]: Get tasks logic is not yet implemented yet, I will complete it this week.

File: ClickUpApp.ts (hunk 13 above)
Comment (mustafahasankhan) [reply ClickUpApp.ts]: I wasting testing it and mistakenly left it public, I implemented getOauth2Config() already below line 102.

File: src/lib/postTask.ts (hunk 7 above)
Comment (mustafahasankhan) [reply src/lib/postTask.ts:68]: Yes sir, it works.

File: src/handlers/ExecuteBlockActionHandler.ts (hunk 3 above)
Comment (mustafahasankhan) [reply src/handlers/ExecuteBlockActionHandler.ts]: It was left so that I remember to implement it next week.

File: src/lib/getTasks.ts (hunk 10 above)
Comment (mustafahasankhan) [reply src/lib/getTasks.ts]: Fixed sir.

File: src/handlers/ExecuteViewSubmitHandler.ts (hunk 11 above)
Comment (mustafahasankhan) [reply src/handlers/ExecuteViewSubmitHandler.ts:21]: Done, sir.

File: src/handlers/ExecuteBlockActionHandler.ts (hunk 2 above)
Comment (mustafahasankhan) [reply src/handlers/ExecuteBlockActionHandler.ts]: Fixed, sir.

File: src/lib/persistence.ts (hunk 4 above)
Comment (mustafahasankhan) [reply src/lib/persistence.ts:4]: Sure sir, I will avoid using `any` wherever possible but what is the alternative way here?

**Review by AllanPazRibeiro — APPROVED** [review]
LGTM
